### PR TITLE
Fix asinh of negative values

### PR DESCRIPTION
**Pull request by Ralith**
Rust's current implementation of asinh has [large errors](https://www.wolframalpha.com/input/?i=arcsinh%28x%29%2C+ln%28x%2B%28x%5E2%2B1%29%5E0.5%29%2C+x+from+-67452095.07139316+to+0) in its negative range. ~These are (mostly) not numerical, but rather seem due to an incorrect implementation.~ This appears to be due to avoidable catastrophic cancellation.
[Playground before/after](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bd04ae6d86d06612e4e389a8b95d19ab).
[glibc uses](https://github.com/bminor/glibc/blob/81dca813cc35f91414731fdd0ff6b756d5e1827f/sysdeps/ieee754/dbl-64/s_asinh.c#L56) abs here.

Many thanks to @danieldeankon for finding this weird behavior, @jebrosen for diagnosing it, and @toasteater for identifying the probable implementation error!

### CODE REVIEW

**Comment by toasteater** [comment]
To think about it after the fact, this probably *is* a numerical issue, though. `self + ((self * self) + 1.0).sqrt()` is supposed to produce a very small value whose `.ln()` is equal to `-(self.abs() + ((self * self) + 1.0).sqrt()).ln()`. In practice, however, once the absolute value of `self` goes above 2^26 (67108864, compare to 67452095 in the original example), `self * self` will be around 2^53, and the `+1.0` will no longer work. That makes `self + ((self * self) + 1.0).sqrt()` dangerously close (or equal to) zero, leading to the precision problems and `-inf`s in the original example.

The `.abs()` addition is a transformation that works because `asinh` is odd, and additions are much more permissive when there are no negative numbers involved.

**Comment by shepmaster** [comment]
I don't have the numerical chops to evaluate this, so I'll be reassigning, but it seems like a really bad idea to "fix" this without (a) even stating what the problem is in the commit message (links to external sites are great, but ultimately very fragile) (b) adding any kind of automated test to prevent regressions.

r? @cramertj

**Comment by Ralith** [comment]
Good points. Added discussion of the numerical issue and rationale for the correctness of the solution to the commit, and threw in a basic regression test.

File: src/libstd/f64.rs [hunk 1]
```diff
@@ -1443,6 +1443,8 @@ mod tests {
         // issue 63271
         assert_approx_eq!(2.0f64.asinh(), 1.443635475178810342493276740273105f64);
         assert_approx_eq!((-2.0f64).asinh(), -1.443635475178810342493276740273105f64);
+        // regression test for the catastrophic cancellation fixed in 72486
+        assert_approx_eq!((-67452098.07139316f64).asinh(), -67452098.07139316f64.asinh());
```
Comment (tspiteri) [inline src/libstd/f64.rs:1447]: I think using the expected exact result is better than another computation. Using the same number of significant digits as the tests above, that would be `-18.72007542627454439398548429400083`.

File: src/libstd/f32.rs [hunk 2]
```diff
@@ -1414,6 +1414,8 @@ mod tests {
         assert!((-0.0f32).asinh().is_sign_negative()); // issue 63271
         assert_approx_eq!(2.0f32.asinh(), 1.443635475178810342493276740273105f32);
         assert_approx_eq!((-2.0f32).asinh(), -1.443635475178810342493276740273105f32);
+        // regression test for the catastrophic cancellation fixed in 72486
+        assert_approx_eq!((-67452098.07139316f64).asinh(), -67452098.07139316f64.asinh());
```
Comment (tspiteri) [inline src/libstd/f32.rs:1418]: This is for `f64`, not for `f32`. 3000 would be enough for `f32`,
```rust
assert_approx_eq!((-3000.0f32).asinh(), -8.699514775987968673236893537700647f32);
```

File: src/libstd/f64.rs (hunk 1 above)
Comment (Ralith) [reply src/libstd/f64.rs]: Applied.

File: src/libstd/f32.rs (hunk 2 above)
Comment (Ralith) [reply src/libstd/f32.rs]: Whoops, good catch, thanks.

**Comment by Ralith** [comment]
Ping?

**Comment by Mark-Simulacrum** [comment]
Hm, I'm not sure who a good reviewer here would be.

Cc @rust-lang/libs - do we have anyone?

**Review by dtolnay — APPROVED** [review]
I can review this kind of thing.

Asinh is odd, and it's clear we're getting better results in the positive range, so I'm happy to take this improvement. [libclc does the abs there too.](https://github.com/llvm/llvm-project/blob/a974b33a10745b528c34f0accbd230b0a4e1fb87/libclc/generic/lib/math/asinh.cl#L58-L60)

```cpp
// Arguments such that 4.0 <= abs(x) <= 1/sqrt(epsilon) are
// approximated by asinhf(x) = ln(abs(x) + sqrt(x*x+1))
// with the sign of x (see Abramowitz and Stegun 4.6.20)
```

Following up on this, it would be good to look at an **appropriately licensed** high quality implementation and consider whether we should pick up what they do on some of the other intervals, namely close to zero and very far from zero where `ln(abs(x) + sqrt(x*x + 1))` still isn't good.

File: src/libstd/f32.rs [hunk 3]
```diff
@@ -835,7 +835,7 @@ impl f32 {
         if self == Self::NEG_INFINITY {
             Self::NEG_INFINITY
         } else {
```
Comment (dtolnay) [inline src/libstd/f32.rs:837]: This branch can be removed, since the other branch now works for -inf.

Comment (Ralith) [reply src/libstd/f32.rs]: Good catch!

**Pull request merged**